community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java

```diff
@@ -960,27 +962,44 @@ public Property nodeSetProperty( KernelStatement state, long nodeId, DefinedProp
         try ( Cursor<NodeItem> cursor = nodeCursorById( state, nodeId ) )
         {
             NodeItem node = cursor.get();
-            Property existingProperty;
-            try ( Cursor<PropertyItem> properties = node.property( property.propertyKeyId() ) )
+            Property existingProperty = readProperty( property.propertyKeyId(), node, EntityType.NODE );
+
+            // TODO Explicitly auto-index properties, even if setting the same value as the existing one.
+            // This because there are established ways of rebuilding auto-indexes which involves this function.
+            autoIndexProperty( nodeId, property, ops, existingProperty, autoIndexing.nodes() );
+
+            if ( !property.equals( existingProperty ) )
```

Comment:
We also have `graphSetProperty` which is not that widely used, as we know, but shouldn't that be updated as well with similar check?